### PR TITLE
TAS: replace nodes only on pod termination; deprecate fixed-time NotReady marking

### PR DESCRIPTION
**Pull request by yakticus**
#### What type of PR is this?

/kind feature
/kind deprecation
/area tas

#### What this PR does / why we need it:

Marking a TAS node as failed after a fixed `NotReady` duration (`NodeFailureDelay`, 30s) is premature in principle: a node can become Ready again after an arbitrary amount of time (cloud node-repair mechanisms are similarly unpredictable), and while the workload's Pods keep running on it, a recomputed topology assignment diverges from the node the Pods actually occupy, corrupting per-node capacity accounting in both directions — the real node is treated as free and over-admitted, while the newly assigned node is blocked for admissions it cannot serve. On our production fleet, transient `NotReady` blips with surviving Pods are far more common than real node deaths.

As proposed by @mimowo in https://github.com/kubernetes-sigs/kueue/issues/12735#issuecomment-4959559208, node replacement is now driven solely by pod termination (previously the `TASReplaceNodeOnPodTermination` behavior, Beta since 0.14). The fixed-time marking is retained behind a new sunset gate:

| `TASReplaceNodeDueToNotReadyOverFixedTime` | `TASReplaceNodeOnPodTermination` | NotReady-node marking trigger |
|---|---|---|
| enabled | enabled | pod termination OR 30s timer, whichever first (pre-0.19 default) |
| enabled | disabled | 30s timer only (pre-0.14 behavior) |
| **disabled (default since 0.19)** | *any* | **pod termination only** |

Deleted nodes and nodes without a `Ready` condition are still marked immediately in all configurations. Gate lifecycle: Beta (default on) at 0.17/0.18 for backports, Deprecated (default off) since 0.19, removal targeted for 0.21.

#### Which issue(s) this PR fixes:

Part of #13042

Related to #11858 — note this PR does **not** fix that issue: the ungater can still pin a replacement pod against a stale `TopologyAssignment` when it observes the pod before the assignment rewrite; changing the marking trigger does not remove that race.

#### Special notes for your reviewer:

- The gate spec must carry both entries on main: the feature-gate machinery rejects a gate born `Deprecated` ("must provide a 1.0 entry indicating initial state"), so the `{0.17 Beta}` entry cannot be deferred to the cherry-picks as we did for #12980.
- While the sunset gate is enabled, the `TASReplaceNodeOnPodTermination` interplay is unchanged (backport-safe); once disabled, that gate has no effect — it is effectively absorbed by the new default and can be deprecated on its own schedule (happy to do that here or in a follow-up, your call).
- The TAS integration suite passes unchanged under the new default — the timer was not load-bearing in any existing spec. New coverage: unit cases for all four gate combinations (including both-disabled, which behaves as termination-only), an integration spec with a real running Pod (node NotReady past the delay → not marked; Pod terminating → hot swap), and a config-validation case for enabling the gate without `TASFailedNodeReplacement`.
- KEP-2724 and the TAS concepts page are updated in this PR.

#### Does this PR introduce a user-facing change?

```release-note
TAS: Added a fix for premature node replacement when a node remains NotReady while the workload's Pods are still running, which could cause the topology assignment to diverge from the actual Pod placement and corrupt per-node capacity accounting. The termination-driven behavior applies when TASReplaceNodeDueToNotReadyOverFixedTime is disabled. The gate is deprecated and disabled by default in 0.19+, and Beta and enabled by default in the 0.17 and 0.18 release branches.
```

### CODE REVIEW

**Comment by mimowo** [comment]
Thank you 👍 cc @tenzen-y @dkaluza ptal

**Comment by tenzen-y** [comment]
> Thank you 👍 cc @tenzen-y @dkaluza ptal

ACK

File: test/integration/singlecluster/tas/tas_test.go [hunk 1]
```diff
@@ -1535,6 +1535,103 @@ var _ = ginkgo.Describe("Topology Aware Scheduling", ginkgo.Ordered, func() {
...
+						NodeName(nodeName).
+						Request(corev1.ResourceCPU, "1").
+						Obj()
+					util.MustCreate(ctx, k8sClient, pod)
+					ginkgo.DeferCleanup(func() {
+						p := &corev1.Pod{}
+						if err := k8sClient.Get(ctx, client.ObjectKeyFromObject(pod), p); err == nil {
+							p.Finalizers = nil
+							_ = k8sClient.Update(ctx, p)
+						}
```
Comment (mimowo) [inline test/integration/singlecluster/tas/tas_test.go:1579]: nit, use controllerutil.RemoveFinalizer(&createdPod, "kueue.x-k8s.io/integration-test")
for better code re-use, also to prevent silencing errors. We may wrap it in gomega.Eventually, similarly as here: https://github.com/kubernetes-sigs/kueue/blob/main/test/integration/singlecluster/controller/jobs/pod/pod_controller_test.go#L353-L357

**Review by mimowo — COMMENTED** [review]
/lgtm
/approve
Thank you for the fix. Such a placement drift can be quite problematic when it happens. Especially, in case of assignments drifting for inference workloads, like LWS, because then we pod running on the old domain may block the capacity for a long time. We observed similar phenomenon as well, cc @dkaluza.

**Comment by mimowo** [comment]
@yakticus let me wait a bit with unholding in case @dkaluza or @tenzen-y have some comments, but in the meanwhile I think you could already prep the cherry-picks. Small adjustments, if any, can be easily ported between the PRs

File: test/integration/singlecluster/tas/tas_test.go [hunk 2]
```diff
@@ -1535,6 +1535,103 @@ var _ = ginkgo.Describe("Topology Aware Scheduling", ginkgo.Ordered, func() {
...
+						g.Expect(wl1.Status.UnhealthyNodes).To(gomega.BeEmpty())
+						g.Expect(wl1.Status.Admission.PodSetAssignments[0].TopologyAssignment).Should(gomega.BeComparableTo(originalAssignment))
+					}, util.ConsistentDuration, util.ShortInterval).Should(gomega.Succeed())
+				})
+
+				ginkgo.By("deleting the pod so it becomes terminating", func() {
+					gomega.Expect(k8sClient.Delete(ctx, pod)).To(gomega.Succeed())
+				})
+
+				ginkgo.By("verify the workload has corrected TopologyAssignment", func() {
```
Comment (mimowo) [inline test/integration/singlecluster/tas/tas_test.go:1611]: Nice additional check to confirm the HotSwap works once the delete happens 👍

File: test/integration/singlecluster/tas/tas_test.go [hunk 3]
```diff
@@ -1535,6 +1535,103 @@ var _ = ginkgo.Describe("Topology Aware Scheduling", ginkgo.Ordered, func() {
...
+					}, util.Timeout, util.Interval).Should(gomega.Succeed())
+				})
+
+				ginkgo.By("removing the pod finalizer", func() {
+					gomega.Eventually(func(g gomega.Gomega) {
+						g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(pod), pod)).To(gomega.Succeed())
+						pod.Finalizers = nil
+						g.Expect(k8sClient.Update(ctx, pod)).To(gomega.Succeed())
+					}, util.Timeout, util.Interval).Should(gomega.Succeed())
+				})
```
Comment (mimowo) [inline test/integration/singlecluster/tas/tas_test.go:1632]: Is that needed since you do it in DeferCleanup?

**Comment by mimowo** [comment]
cc @j-skiba

**Comment by mimowo** [comment]
Ups, I thought the PR is on-hold, @yakticus please address the comments in a follow up.

File: pkg/controller/tas/node_controller.go [hunk 4]
```diff
@@ -154,7 +154,7 @@ func (r *nodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.
 		timeSinceNotReady := r.clock.Now().Sub(readyCondition.LastTransitionTime.Time)
 		remainingTime := NodeFailureDelay - timeSinceNotReady
 		timerExpired = remainingTime <= 0
-		if !timerExpired && !features.Enabled(features.TASReplaceNodeOnPodTermination) {
+		if features.Enabled(features.TASReplaceNodeDueToNotReadyOverFixedTime) && !timerExpired && !features.Enabled(features.TASReplaceNodeOnPodTermination) {
```
Comment (j-skiba) [inline pkg/controller/tas/node_controller.go:157]: nit

```suggestion
		if !timerExpired && !replaceOnPodTermination() {
```

File: pkg/controller/tas/node_controller.go [hunk 5]
```diff
@@ -411,6 +411,13 @@ func (r *nodeReconciler) getWorkloadStatus(
 	}
 }
 
+// replaceOnPodTermination reports whether node replacement is driven by pod
+// termination. Once the fixed-time marking is sunset (gate disabled), this is
+// unconditionally true and TASReplaceNodeOnPodTermination has no effect.
+func replaceOnPodTermination() bool {
+	return features.Enabled(features.TASReplaceNodeOnPodTermination) || !features.Enabled(features.TASReplaceNodeDueToNotReadyOverFixedTime)
```
Comment (mimowo) [inline pkg/controller/tas/node_controller.go:418]:
```suggestion
	return features.Enabled(features.TASReplaceNodeOnPodTermination)
```
probably a corner case, but let's avoid this tricky implicit behavior.

File: test/integration/singlecluster/tas/tas_test.go (hunk 1 above)
Comment (yakticus) [reply test/integration/singlecluster/tas/tas_test.go:1579]: Done in #13095 (both cleanup sites).

File: test/integration/singlecluster/tas/tas_test.go (hunk 3 above)
Comment (yakticus) [reply test/integration/singlecluster/tas/tas_test.go:1632]: Kept, with a note in #13095: `DeferCleanup` runs after the suite-level `AfterEach`, so without the in-spec step the namespace teardown waits out the pod finalizer. The `DeferCleanup` remains as the backstop when the spec fails before reaching it.

File: pkg/controller/tas/node_controller.go (hunk 4 above)
Comment (yakticus) [reply pkg/controller/tas/node_controller.go:157]: Done in #13095.

File: pkg/controller/tas/node_controller.go (hunk 5 above)
Comment (yakticus) [reply pkg/controller/tas/node_controller.go:418]: Done in #13095 — the helper now checks only `TASReplaceNodeOnPodTermination`; unit cases and docs/KEP wording updated to match (disabling it retains the fixed-time marking).

**Comment by yakticus** [comment]
> Ups, I thought the PR is on-hold, @yakticus please address the comments in a follow up.

@mimowo see #13095

File: pkg/controller/tas/node_controller.go (hunk 5 above)
Comment (tenzen-y) [reply pkg/controller/tas/node_controller.go:418]: +1
I actually suspected that this helper has bug-ish behavior.
I will check it deeply later.

**Comment by mimowo** [comment]
/remove-kind feature
/kind bug
Actually, this we are deprecating the mode because we concluded it is buggy.

**Comment by mimowo** [comment]
/release-note-edit
```release-note
TAS: Fixed a bug where a node remaining `NotReady` for 30 seconds could trigger premature node replacement while the workload's Pods were still running, causing the topology assignment to diverge from the actual Pod placement and resulting in incorrect per-node capacity accounting. Node replacement is now triggered when the affected Pods are terminating or terminated. The previous fixed-time replacement behavior is guarded by the `TASReplaceNodeDueToNotReadyOverFixedTime` feature gate, which is deprecated and disabled by default in 0.19+, and Beta and enabled by default in the 0.17 and 0.18 release branches.
```

**Comment by mimowo** [comment]
/remove-kind deprecation
/kind bug
It is a deprecation, but we rarely use the mode. This is also a bugfix. Also, we detail the deprecation steps in the release note anyway.

**Comment by mimowo** [comment]
/release-note-edit
```release-note
TAS: Added a fix for premature node replacement when a node remains NotReady while the workload's Pods are still running, which could cause the topology assignment to diverge from the actual Pod placement and corrupt per-node capacity accounting. The termination-driven behavior applies when TASReplaceNodeDueToNotReadyOverFixedTime is disabled. The gate is deprecated and disabled by default in 0.19+, and Beta and enabled by default in the 0.17 and 0.18 release branches.
```
